### DISCUSSION TITLE
Request and Response Interception in Scalar Client

### DISCUSSION THREAD

**Discussion by rinkusam12** [scalar/scalar#2810] [Ideas]
Implement a mechanism to intercept requests and responses in the Scalar client to enable custom operations such as adding headers and modifying responses. This feature is similar to Swagger UI's requestInterceptor and responseInterceptor methods.

**Feature Details:**
- **Request Interceptor:** A method to intercept outgoing requests, allowing users to add custom headers. For example, after a successful login, the interceptor should be able to automatically include an authentication token in the headers of subsequent requests.
- **Response Interceptor:** A method to intercept incoming responses, enabling users to perform operations such as storing tokens or modifying the response data before it is handled by the application.

**Use Case:**
- On a successful login response, the response interceptor should extract the authentication token and store it in local storage.
- On subsequent requests, the request interceptor should retrieve the stored token and include it in the request headers to maintain authenticated sessions. 

If it possible to have something similar to Postman where we can store something from response to environment variables and then use it for request that would be even more better improving the already refined UX to a even better UX.

**Comment by hanspagel** [comment]
Love it! Actually, I wrote a way to intercept requests already, but it’s only living on my hard disk. Will put together a PR.

**Comment by floxay** [comment]
Any news on this?
Request interceptor(s) to handle CSRF would be amazing.

**Comment by lgcmotta** [comment]
@hanspagel, any news on this? I want to know if it's possible to intercept the scalar UI request to the API using the Scalar.AspNetCore package. I have secured access to the Scalar UI using OIDC but, the problem is that the browser keeps sending the cookies with the API requests. In Swagger UI, I was able to inject an interceptor function in JS to decide if the cookies should be added or not.

**Comment by JasonLovesDoggo** [comment]
bump! Currently, i'm not able to use scalar as backend requires the CSRF token to be sent as a header and without a request interceptor, I cannot add that logic.

**Comment by doug-shontz** [comment]
Bump on this as well. This is the one feature keeping me on Swagger.

**Comment by kiwiant** [comment]
Looking at https://guides.scalar.com/scalar/scalar-api-references/configuration I used the **fetch** method to override the document request to add an HTTP header but it also looks like you can do the same for requests with the **onBeforeRequest** and **onRequestSent** events.
```
{
  onBeforeRequest: ({ request }) => {
    // Add a custom header to all requests
    request.headers.set('X-Custom-Header', 'test')
  }
}
```

**Reply by lgcmotta** [reply]
> Looking at https://guides.scalar.com/scalar/scalar-api-references/configuration I used the **fetch** method to override the document request to add an HTTP header but it also looks like you can do the same for requests with the **onBeforeRequest** and **onRequestSent** events.
> 
> ```
> {
>   onBeforeRequest: ({ request }) => {
>     // Add a custom header to all requests
>     request.headers.set('X-Custom-Header', 'test')
>   }
> }
> ```

This might work... I was looking at the [.NET integration](https://guides.scalar.com/scalar/scalar-api-references/integrations/net-aspnet-core/integration#scalar-api-reference-for-net-aspnet-core__configuration-options__custom-javascript-configuration) documentation, and I think we can inject this JS code using the `WithJavaScriptConfiguration` extension.

**Comment by JeroMiya** [comment]
For those looking for a solution like I was, @lgcmotta was correct, except the `onBeforeRequest` callback needs to be defined slightly differently. It can be done like so:

In your setup code:
```CSharp
            app.MapScalarApiReference(options =>
            {
                options.WithJavaScriptConfiguration("./scalar-config.js");
            });
```

And `scalar-config.js` is defined like so, using `requestBuilder` instead of `request`, as the `request` is considered immutable here:
```JS
export default {
    onBeforeRequest: ({ requestBuilder }) => {
        requestBuilder.headers.set('X-CSRF', '1')
    }
};
```

There's actually an example much like this in the docs: https://scalar.com/products/api-references/configuration#onbeforerequest

As a side note, it would be nice if we could define the JS config module inline in the C# code, so it can more easily be excluded from the final app build, where Scalar might not be enabled (I had to add a conditional `<None Include="etc.." />` element in the `.csproj` file).

**Comment by hanspagel** [comment]
https://github.com/scalar/scalar/pull/10102 merged, gonna go out soon
